### PR TITLE
Aggregate transactions by account holder

### PR DESCRIPTION
**Pull request by RikerlWien**
- extended EntityConnexion class to aggregate transactions by account holder
- some refactoring/documenting of the EntityConnexion class
- add two example scripts demonstrating how to build transaction graphs for a) an account b) a account holder.

### CODE REVIEW

File: codebase/entity_connexion.py [hunk 1]
```diff
@@ -90,65 +150,70 @@ def get_transactions(self):
...
 
-        df = self.transactions
+        transaction_table = self.transactions
         this_node = self.entity_id
 
         # Sometimes there are missing values in the transaction dataframe
-        for val in {'transferring', 'acquiring'}:
-            if df[f'{val}Entity_id'].isnull().any():
-                warnings.warn(f'  Some {val}Entity IDs missing ... replacing')
+        for side in {'transferring', 'acquiring'}:
```
Comment (lmillischer) [inline codebase/entity_connexion.py]: pourquoi pas self.sides?

File: codebase/entity_connexion.py [hunk 2]
```diff
@@ -1,83 +1,143 @@
...
 
         transactions = transactions[column_list].copy()
 
         if self.entity_type == 'Account':
-            for v in ['transferring', 'acquiring']:
-                transactions.rename(columns={f'{v}Account_id': f'{v}Entity_id',
-                                             f'{v}AccountName': f'{v}Entity_name',
-                                             f'{v}AccountType': f'{v}Entity_type'}, inplace=True)
+            # rename AccountHolder -> Entity
+            for side in ['transferring', 'acquiring']:
```
Comment (lmillischer) [inline codebase/entity_connexion.py]: pourquoi pas self.sides?

File: codebase/entity_connexion.py [hunk 3]
```diff
@@ -1,83 +1,143 @@
...
-                warnings.warn('Columns not available. Investigate!')
-                return
+
+        if not set(column_list).issubset(transactions.columns):
+            warnings.warn('At least one of the required columns is missing from the transaction table. Investigate!')
+            return
 
         transactions = transactions[column_list].copy()
 
         if self.entity_type == 'Account':
```
Comment (lmillischer) [inline codebase/entity_connexion.py]: pourquoi pas EntityType.Account?

File: codebase/entity_connexion.py [hunk 4]
```diff
@@ -1,83 +1,143 @@
...
-            for v in ['transferring', 'acquiring']:
-                transactions.rename(columns={f'{v}Account_id': f'{v}Entity_id',
-                                             f'{v}AccountName': f'{v}Entity_name',
-                                             f'{v}AccountType': f'{v}Entity_type'}, inplace=True)
+            # rename AccountHolder -> Entity
+            for side in ['transferring', 'acquiring']:
+                transactions.rename(columns={f'{side}Account_id': f'{side}Entity_id',
+                                             f'{side}AccountName': f'{side}Entity_name',
+                                             f'{side}AccountType': f'{side}Entity_type'}, inplace=True)
         elif self.entity_type == 'AccountHolder':
```
Comment (lmillischer) [inline codebase/entity_connexion.py]: pourquoi pas EntityType.AccountHolder?

File: codebase/entity_connexion.py [hunk 5]
```diff
@@ -1,83 +1,143 @@
...
-            #  then rename columns, drop transactions from accountholder_id to itself
-            pass
+            # map each (transferring or acquiring) account to its account holder
+            for side in self.sides:
+                account_to_accountHolder_temp = account_to_accountHolder.copy()
+                account_to_accountHolder_temp.rename(columns={"account_id": f"{side}Account_id",
+                                                              "accountHolder_id": f"{side}AccountHolder_id",
+                                                              "accountHolder_name": f"{side}AccountHolder_name"}, inplace=True)
+                transactions = transactions.merge(account_to_accountHolder_temp)
+
```
Comment (lmillischer) [inline codebase/entity_connexion.py:140]: I would always add a test that all has been merged well:

`merge(...., indicator=True)`

Also only merge left:

`merge(..., how='left')

Then check:

`merge_test = df.groupby('_merge', asindex=False)['_merge'].count()`
`if len(merge_test[merge_test._merge != 'both'])>1 ValueError`

File: codebase/entity_connexion.py [hunk 6]
```diff
@@ -1,83 +1,143 @@
...
+    Build a pandas DataFrame where each row corresponds to an Account. Each entry contains:
+    - the account id (account_id)
+    - the id of its account holder (accountHolder_id)
+    - the name of its account holder name (accountHolder_name)
+
+    The DataFrame is used to map each account with its account holder. Since this mapping is independent of a
+    particular Entity under consideration, we define it at the module level. It is built when the module is imported.
+    """
+    mapping = pd.DataFrame(session.query(Account.id, Account.accountHolder_id).all())
```
Comment (lmillischer) [inline codebase/entity_connexion.py]: mapping_query = session.query(Account.id, Account.accountHolder_id)
    mapping = pd.read_sql(mapping_query.statement, con=session.bind)

File: codebase/entity_connexion.py [hunk 7]
```diff
@@ -90,65 +150,70 @@ def get_transactions(self):
...
+                #  -> is there such a thing as a "type" attribute for AccountHolders of Companies ?
+                # attrs[node]['type'] = df[df['transferringEntity_id'] == node].iloc[0]['transferringEntity_type']
             elif node in trans_entities:
                 attrs[node]['trader_type'] = 'sender'
                 attrs[node]['color'] = color_legend[attrs[node]['trader_type']]
-                attrs[node]['type'] = df[df['transferringEntity_id'] == node].iloc[0]['transferringEntity_type']
-                attrs[node]['name'] = df[df['transferringEntity_id'] == node].iloc[0]['transferringEntity_name']
-                attrs[node]['id'] = df[df['transferringEntity_id'] == node].iloc[0]['transferringEntity_id']
+                # TODO: commented out the line below because there is no "Type" column in the AccountHolder table.
+                #  -> is there such a thing as a "type" attribute for AccountHolders of Companies ?
```
Comment (lmillischer) [inline codebase/entity_connexion.py]: No just for accounts but that we would like to use for the colors, no? I would keep it.

Comment (RikerlWien) [reply codebase/entity_connexion.py]: Ok, defined (empty) fields `acquiringEntity_type` and `transferringEntity_type` for `AccountHolders`.

File: codebase/entity_connexion.py (hunk 5 above)
Comment (RikerlWien) [reply codebase/entity_connexion.py:140]: ok, check implémenté

**Pull request merged**